File: 算法/递归.py
# 递归实现 从 i 开始倒计时
def count(i):
    print(i)  # 我们想实现的功能

    if i == 0:  # 基线条件：停止条件
        return
    else:  # 递归条件：自己调用自己的条件
        count(i - 1)


# 递归数组求数组元素的和
def sum(arr: list):
    if not arr:
        return 0
    else:
        return arr.pop(0) + sum(arr)


# 递归数组得数组元素个数
def count(arr: list):
    if arr:
        return 1 + count(arr[:-1])
    else:
        return 0


# 使用递归找出数组中最大的数
def find_biggest(arr: list):
    if len(arr) == 1:
        return arr[0]
    else:
        # return max(arr[0], find_biggest(arr[1:]))
        a = arr.pop(0)
        b = find_biggest(arr)
        return a if a >= b else b


if __name__ == '__main__':
    # count(10)
    ls = [12, 1, 12, 2]
    # print(sum(ls))
    print(find_biggest(ls))

File: 算法/递归.cpp
//
// Created by Momo on 2023/7/1.
//
#include <iostream>

namespace recursion {
    void count(int i) {
        std::cout << "counting " << i << std::endl;
        if (i == 1)
            return;
        else
            count(i - 1);
    }

    // 利用递归遍历数组并求元素和
    int sum(const int *head, const int *end) {
        if (head == end)
            return 0;
        else
            return *head + sum(head + 1, end);
    }


    // 使用递归数组统计数组元素的个数
    int count_array(const int *head, const int *end) {
        if (head == end)
            return 0;
        else
            return 1 + count_array(head + 1, end);
    }

    // 使用递归找出数组中的最大值
    int find_max(const int *head, const int *end)
    {
      if (head == end)
          return *head;
      else
          return std::max(find_max(head+1, end), *head);
    }

}

int main() {
//    count(5);
    int arr[] = {20, 12, 6};

//    std::cout << sum(arr, arr + 3) << std::endl;
    // 使用统计递归数组
//    std::cout << recursion::count_array(arr, arr + 3) << std::endl;
    std::cout << recursion::find_max(arr, arr + 3) << std::endl;
    return 0;
}